opal/mca/threads/qthreads/threads_qthreads_module.c

```diff
@@ -81,6 +81,11 @@ int opal_thread_start(opal_thread_t *t)
 
 OBJ_CLASS_DECLARATION(opal_thread_t);
 
+int opal_tsd_key_create_tracked(opal_tsd_key_t *key, opal_tsd_destructor_t destructor)
```

Comment:
@hppritcha is there a way to implement this functionality with qthreads?

Reply:
pinging the qthreads and argobots crews

Reply:
Any documentation on the tracked keys anywhere? @hppritcha @artpol84

Reply:
@janjust Could you please apply this patch? It will fix the implementation in the Argobots module.

```diff
diff --git a/opal/mca/threads/argobots/threads_argobots_module.c b/opal/mca/threads/argobots/threads_argobots_module.c
index 106549a..f166698 100644
--- a/opal/mca/threads/argobots/threads_argobots_module.c
+++ b/opal/mca/threads/argobots/threads_argobots_module.c
@@ -141,7 +141,10 @@ int opal_tsd_key_create_tracked(opal_tsd_key_t *key, opal_tsd_destructor_t destr
 
 int opal_tsd_key_create(opal_tsd_key_t *key, opal_tsd_destructor_t destructor)
 {
-    return OPAL_ERR_NOT_IMPLEMENTED;
+    opal_threads_argobots_ensure_init();
+    int rc;
+    rc = ABT_key_create(destructor, key);
+    return (ABT_SUCCESS == rc) ? OPAL_SUCCESS : OPAL_ERROR;
 }
 
 int opal_tsd_keys_destruct(void)
```

Reply:
@shintaro-iwasaki what is the counterpart for this? the analog of pthread_key_delete?

Reply:
Sorry for the late reply. Yes, it is. This patch implements "`OPAL_ERR_NOT_IMPLEMENTED`" in `opal_tsd_key_create()` of the Argobots implementation.

Reply:
@shintaro-iwasaki done, but just fyi, the _tracked version is left as unimplemented.

Reply:
@janjust The description of https://github.com/open-mpi/ompi/pull/7877/commits/a4ce523d51ce4525687e308b046a05a84a3ca352 is confusing. How can it implement `OPAL_ERR_NOT_IMPLEMENTED`?

Reply:
It's a copy paste of @shintaro-iwasaki description, since it's his patch that was applied. But I can change it - I'm easy :)

Reply:
I would like to clarify this situation:
- There are three MCA thread modules: Pthreads, Qthreads, and Argobots
- The first commit by @janjust only implemented Pthreads. The Argobots/Qthreads counterparts were marked as `OPAL_ERR_NOT_IMPLEMENTED`.
- I created a patch that fixes the "Argobots" counterpart. (`opal/mca/threads/argobots/threads_argobots_module.c`)
- The Qthreads counterpart is still marked as `OPAL_ERR_NOT_IMPLEMENTED` because the Qthreads module has not implemented TSD yet (`opal/mca/threads/qthreads/threads_qthreads_module.c`).

Reply:
@janjust If you don't mind, please squash-merge my patch. This just complements your original commit.

Reply:
@shintaro-iwasaki done

Reply:
@janjust Thank you very much!

Reply:
Looking at the way TSDs are used in the UCX integration I wonder if the patch provided by @shintaro-iwasaki really does what we want/expect. Provided I build my Open MPI installation with Argobots support and use tasks that are mapped to Argobots (using BOLT for example) and then issue only one or two MPI calls in each task, the thread-specific data has to be created over and over again. It seems that `_tlocal_add_ctx` and `_tlocal_ctx_connect` are pretty expensive and that would have to be done for *each task*, which seems like a prohibitively large overhead. @shintaro-iwasaki please correct me if I'm wrong here. From a UCX standpoint, pthread-based TSDs should be sufficient, no?

Reply:
@janjust I'd prefer to split our changes and changes provided by @shintaro-iwasaki so the contribution is properly distributed.

Reply:
@devreal Thanks! I am not fully sure how UCX uses TLS, but it could be expensive if all tasks/threads create UCX contexts as you wrote.
In terms of correctness, I do not know if tasks can safely share a UCX context that is associated with Pthreads. For example, if a UCX context has a state and each "thread" that owns it needs to remember its state, yielding a task after changing the state can cause a problem.
If there is a scheduling limitation it might work, but this is not trivial to me so far. In any case, this should be a performance problem that needs to be solved.

Reply:
Thanks, I believe this should be investigated separately. I agree with @artpol84 that this patch should be separated out into its own PR as this is an orthogonal issue to the proposal here.